confmap/confmap.go

```diff
@@ -310,33 +315,47 @@ func unmarshalerEmbeddedStructsHookFunc() mapstructure.DecodeHookFuncValue {
 
 // Provides a mechanism for individual structs to define their own unmarshal logic,
 // by implementing the Unmarshaler interface.
-func unmarshalerHookFunc(result any) mapstructure.DecodeHookFuncValue {
+func unmarshalerHookFunc(result any, allowTopLevelUnmarshaler bool) mapstructure.DecodeHookFuncValue {
 	return func(from reflect.Value, to reflect.Value) (any, error) {
 		if !to.CanAddr() {
 			return from.Interface(), nil
 		}
 
 		toPtr := to.Addr().Interface()
 		// Need to ignore the top structure to avoid circular dependency.
-		if toPtr == result {
+		if toPtr == result && !allowTopLevelUnmarshaler {
 			return from.Interface(), nil
 		}
 
 		unmarshaler, ok := toPtr.(Unmarshaler)
 		if !ok {
 			return from.Interface(), nil
 		}
-
-		if _, ok = from.Interface().(map[string]any); !ok {
-			return from.Interface(), nil
+		unmarshalMethod := reflect.ValueOf(toPtr).MethodByName("Unmarshal")
+		// check that the Unmarshal method is not defined on one of the squash embedded structs defined.
+		// This allows us to have composition where the Unmarshaler may embed structs implementing Unmarshaler.
+		if to.Type().Kind() == reflect.Struct {
+			for i := 0; i < to.Type().NumField(); i++ {
+				// check embedded structs:
+				f := to.Type().Field(i)
+				if f.IsExported() && f.Anonymous && slices.Contains(strings.Split(f.Tag.Get("mapstructure"), ","), "squash") {
```

Comment:
Why do we need to check for the `mapstructure` tag?

Reply:
This is explained in the comment above: if the embedded struct is using the `,squash` value for the `mapstructure` key, then this struct Unmarshal method should not be used as the main struct unmarshaling.